docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md

```diff
@@ -0,0 +1,288 @@
+neighbors may stil occurs during runtime; on the scheduling part. we need some history based optimization during 
+autoscale and allocation for cost-efficiency.
+
+The optimization above relies on the analysis of historical metric data for prediction and detection. So we propose a general 
+forecasting framework, retrieving metric data and build models such as resource usage and runtime quality. These models
+can be used on Interference Detection for higher QoS, and Resource Prediction to improve resource efficiency. 
+
+## Motivation
+
+Several topics are still waiting to be proposed in Koordinator yet!
+
+- Interference Detection: Koordlet has provided several QoS Strategies to mitigate the noisy neighbor problem: setting different resource limit
+for pods for different QoS Class. These are pre-active and preventive measures, trying to prevent noisy neighbor happens.
+We need a proactive approach by collecting metrics and analysing periodically and continually, catching the bad guy once
+someone got interference by noisy neighbors.
+
+- Resource Prediction: Koordinator needs a better estimation of resource consumption during the resource overcommitment.
+Since a radical strategy may cause resource competition, and a conservative leads to the waste of resource. This recommendation
+also supports the new [MPA framework](https://github.com/kubernetes/autoscaler/blob/master/multidimensional-pod-autoscaler/AEP.md),
+which allows integrating with extended Recommenders.
+
+- Forecasting of self-defined metrics: Forecasting is more than just for Kubernetes objects like pods and containers.
+Actually, metric APIs already supports self-defined types of data, e.g. `custom.metrics.k8s.io`, `external.metrics.k8s.io`, and
+the commonly used `Prometheus`. In a two-level scheduling architecture like FaaS, tasks running in containers varies over
+time, so prediction of task metric is valuable when routing task from central.
+
+These topics are all related to data analysis, so we need a general framework for prediction and detection to support 
+different scenarios. 
+
+### Goals
+- Design a holistic framework with modules of profile control, forecasting and data scraping to support data prediction and detection.
+
+- Clarify the data type and transfer links, and how to save the storage and collection overhead.
+
+- Collaboration and division of builtin and self-extended metric exporters, and adaption of multiple using scenarios.
+
+### Non-Goals
+- Design of forecasting algorithm. The interference detection and resource prediction may share same prediction models,
+however, this is over-detailed now. 
+
+- Supports of MPA framework proposed in Kubernetes autoscaler. Although out work can run as a recommender for MPA after
+some protocol conversion works, we can discuss in following proposals.
+
+- Long-term optimization, even though we have already mentioned some ideas about saving transfer cost and storage overhead.  
```

Comment:
May be the storage implement of metrics has a variety of options? For example, InfluxDB. Default is `in memory`.

Reply:
Yes It is, and especially for some algorithms depends on offline training. At the beginning, saving in memory is enough for some plain models.